src/odemis/driver/ueye.py

```diff
@@ -695,9 +695,19 @@ def __init__(self, name, role, device=None, **kwargs):
             # Set the format
             # TODO: color mode based on the maximum BPP
             # cf is_DeviceFeature( IS_DEVICE_FEATURE_CMD_GET_SUPPORTED_SENSOR_BIT_DEPTHS)
-            self._dll.is_SetColorMode(self._hcam, CM_MONO8)
-            self._dtype = numpy.uint8
-            self._metadata[model.MD_BPP] = 8
+            # Loop through possible bit depths and pick the highest one that does not raise an error
+            for bpp, bit_depth in zip([16, 12, 10, 8], [CM_MONO16, CM_MONO12, CM_MONO10, CM_MONO8]):
+                try:
+                    self._dll.is_SetColorMode(self._hcam, bit_depth)
+                    self._dtype = numpy.uint8 if bpp == 8 else numpy.uint16
+                    self._metadata[model.MD_BPP] = bpp
+                    logging.info("Set color mode to {} bits.".format(bpp))
+                    break
+                except UEyeError as err:
+                    if err.errno != 174 or (err.errno == 174 and bpp == 8):  # INVALID_COLOR_FORMAT
+                        raise
+                    else:
+                        logging.info("Set color mode to {} bits not possible, trying a lower value.".format(bpp))
```

Comment:
To make it clearer that the exception is coming from `SetColorMode()`, I'd suggest to put only that call inside the try/except block. Moreover I'd suggest to invert the if, to make it also more readable.
Finally, you can omit the zip, and directly write as it'd be generated.
```suggestion
            for bpp, bit_depth in ((16, CM_MONO16), (12, CM_MONO12), (10, CM_MONO10), (8, CM_MONO8)):
                try:
                    self._dll.is_SetColorMode(self._hcam, bit_depth)
                    break  # found a good color mode
                except UEyeError as err:
                    if err.errno == 174 and bpp > 8:  # INVALID_COLOR_FORMAT
                        logging.debug("Set color mode to {} bits not possible, trying a lower value.".format(bpp))
                    else:
                        raise
           self._dtype = numpy.uint16 if bpp > 8 else numpy.uint8 
           self._metadata[model.MD_BPP] = bpp
           logging.info("Set color mode to {} bits.".format(bpp))
```